First lab reviewing

MaEgV opened the pull request.

besperspektivnyak committed:
add first realization

first (draft) realization of Bacteria and GodSimulator classes

besperspektivnyak committed:
Delete README.md

besperspektivnyak committed:
fix errors

fix minor errors in first version

MaEgV committed:
Typos fix

Resolving conflicts with other classes. Adding brackets

MaEgV committed:
Create population.py

besperspektivnyak committed:
fix logical errors

besperspektivnyak committed:
remove redundant strings

MaEgV committed:
Смена базовой библиотеки

Хранение и отрисовка осуществляется с помощью библиотеки Igraph

besperspektivnyak committed:
добавлены новые и обновленные файлы

изменена структура бактерии, усложнены принципы, по которым наследуются свойства

besperspektivnyak committed:
правка второй версии

добавлены возвращаемые значения, изменены некоторые имена методов на более корректные, константы вынесены как отдельные поля класса

MaEgV committed:
Фикс незначительной логической ошибки

Особи с нулём оставшихся итераций быть не может

MaEgV committed:
Опечатки

besperspektivnyak committed:
добавлены комментарии

добавлены комментарии к классам и методам, изменен метода create_bacetria

besperspektivnyak committed:
удален GodSimulator.py

GodSimulator.py был заменен на EvolutionSimulator.py

MaEgV committed:
Typos fix

MaEgV committed:
Оптимизирован цикл в Population. Дабавлен пример запуска

besperspektivnyak committed:
удален poipulation.py

Старая ненужная версия

besperspektivnyak committed:
Merge pull request #3 from MaEgV/bacteria

Bacteria

MaEgV committed:
Добавлены комментарии. Исправлены логические ошибки в EvolutionSimulator

Теперь вероятность смерти бактерий не монотонно возрастает с новым поколением

MaEgV committed:
Для рефакторинга

MaEgV committed:
Для рефакторинга

MaEgV committed:
Для рефакторинга

MaEgV committed:
Для рефакторинга

MaEgV committed:
Для рефакторинга

MaEgV committed:
Для рефакторинга

MaEgV committed:
Файл остался без изменений

MaEgV committed:
Файл остался без изменений

MaEgV committed:
Изменена схема мутаций

Добавлен абстрактный класс, позволяющий сторонним модулям реализовывать законы, по которым будут происходить генетические изменения в популяции

MaEgV committed:
Изменение системы принятия решений

Добавлено поле с реализацией систем мутаций генома в статическом классе. Пользователи могут менять его значение для контроля над популяцией. Добавлена связь с классом population у бактерий. Это позволяет учитывать внешние факторы при вынесении решения о гибели или размножении

MaEgV committed:
Итератор вынесен в отдельный класс

Итератор вынесен в отдельный класс для возможности повторять эксперимент без создания нового экземпляра класса Population

MaEgV committed:
Добавлены параметры для популяции

Параметры популяции позволяют эволюции корреткировать свои решения
antogonism - степень враждебности особей друг к другу из (0,1)
overpopulation - степень перенаселенности из (0,1)

MaEgV committed:
Тестовый пример для запуска программы

besperspektivnyak committed:
добавлен файл с нужным расширением

besperspektivnyak committed:
Удален файл без расширения

MaEgV committed:
Комментирование класса Itetator

MaEgV committed:
Комментирование класса Population

MaEgV committed:
Комментирование файла Population.parameters

MaEgV committed:
Формирование пакета

MaEgV committed:
Формирование пакета

MaEgV committed:
Add files via upload

MaEgV committed:
Создание README

Первый вариант README для проекта

MaEgV committed:
Исправление README

MaEgV committed:
Исправление примера кода в README

MaEgV committed:
Исправление примера кода в README

MaEgV committed:
Update README.md

MaEgV committed:
Update README.md

MaEgV committed:
Исправление структуры проекта

besperspektivnyak committed:
Удален избыточный файл

iterator.py больше не нужен

besperspektivnyak committed:
Добавлены ключи

Добавлены ключи из iterator.py

besperspektivnyak committed:
Добавлены файлы с тестами

MaEgV committed:
Merge pull request #4 from MaEgV/tests

Добавлены файлы с тестами

besperspektivnyak committed:
Обновлен python-package.yml

Изменено name и добавлены действия при pull requests

besperspektivnyak committed:
Добавлена ветка population

besperspektivnyak committed:
Обновлен requirements.txt

Добавлены новые зависимости

MaEgV committed:
Исправление структуры проекта

MaEgV committed:
Исправление структуры проекта

MaEgV committed:
Исправление структуры проекта

MaEgV committed:
Merge remote-tracking branch 'origin/population' into population

MaEgV committed:
Добавление файла __init__.py

Добавлен файл с инициализацией пакета

MaEgV committed:
Update README.md

MaEgV committed:
Update README.md

MaEgV committed:
Update README.md

MaEgV committed:
Update README.md

MaEgV committed:
Update README.md

MaEgV committed:
Исправление структуры проекта

MaEgV committed:
Исправление структуры проекта

MaEgV committed:
Merge remote-tracking branch 'origin/population' into population

# Conflicts:
#	src/population_modeling/__init__.py

MaEgV committed:
Опечатка в тесте

MaEgV committed:
Опечатка в тестах

MaEgV committed:
Настройка тестов

MaEgV committed:
Исправления в коде по первой части review

Исправления от 09.03 без претензий к reviewer-у

besperspektivnyak committed:
Исправлены структурные недочеты

MaEgV committed:
Фиксы ридми после review

MaEgV committed:
Исправление опечаток

besperspektivnyak committed:
Отладка тестов

besperspektivnyak committed:
Merge remote-tracking branch 'origin/population' into population

besperspektivnyak committed:
Отладка импортов в тестах

besperspektivnyak committed:
Исправление опечаток

besperspektivnyak committed:
Исправление опечаток с импортами

MaEgV committed:
Настройка тестов, опечатки в ридми

besperspektivnyak committed:
Еще однп попытка запустить тесты

besperspektivnyak committed:
Merge remote-tracking branch 'origin/population' into population

# Conflicts:
#	tests/bacteria_iteration_test.py
#	tests/population_iterator_test.py

MaEgV committed:
Опечатки в тестах

MaEgV committed:
Опечатки в тестах

MaEgV committed:
Опечатки в тестах

besperspektivnyak committed:
Еще одна попытка запустить тесты

besperspektivnyak committed:
Merge remote-tracking branch 'origin/population' into population

# Conflicts:
#	tests/bacteria_iteration_test.py
#	tests/population_iterator_test.py

besperspektivnyak committed:
Добавлен файл __init__.py

besperspektivnyak committed:
Исправлен импорт

besperspektivnyak committed:
add mypy check

besperspektivnyak committed:
fix mypy check

besperspektivnyak committed:
fix mypy check

MaEgV committed:
Fix mypy errors

Added ignoring of the third-party library

JDima commented on population_modeling/genome.py:
```diff
@@ -0,0 +1,63 @@
+    p_for_reproduction : float
+        Probability of reproduction
+    """
+    def __init__(self, max_life_time_: int, p_for_death_: float, p_for_reproduction_: float):
```
можно без нижних подчеркиваний, питон это съест и это будет лучше с комментариями коррелировать

JDima commented on README.md:
```diff
@@ -1 +1,58 @@
+
+
+Класс Population имеет 2 публичных метода:
+1. iteration - один жизненный цикл всех бактерий в популяции;
```
так довольно непонятно что именно по функционалу предлагается - лучше снабжать код сниппетами

JDima commented on population_modeling/exceptions.py:
```diff
@@ -0,0 +1,8 @@
+class DeadBacteria(Exception):
+    def __init__(self, bacteria, message='Addressing for a dead bacteria'):
+        super().__init__(message)
+        self.bacteria = bacteria  # here will be id
```
will be лучше помечать как TODO

JDima commented on population_modeling/bacteria.py:
```diff
@@ -0,0 +1,65 @@
+
+class Bacteria:
+    """
+       Represents bacteria with its parameters and ability for reproduction.
```
При комментировании кода можно генерить документацию - вот, например, пакет для этого https://pypi.org/project/Sphinx/ Поппробуйте и положите в папку docs

MaEgV committed:
Intentional bug for mypy

JDima commented on population_modeling/population.py:
```diff
@@ -0,0 +1,118 @@
+
+    def __init__(self, selector_: Selector, max_life_time=5, p_for_death=0.5, p_for_reproduction=0.5):
+        self.selector = selector_
+        self.graph = create_graph(create_bacteria(max_life_time, p_for_death, p_for_reproduction))
```
такое ощущение, что переменная граф не очень информативна с учетом, что это аттрибут класса популяция

JDima commented on population_modeling/population.py:
```diff
@@ -0,0 +1,118 @@
+    INDIVIDUAL_KEY = 'bacteria'
+    GENERATION_KEY = 'generation'
+
+    def __init__(self, selector_: Selector, max_life_time=5, p_for_death=0.5, p_for_reproduction=0.5):
```
` max_life_time=5, p_for_death=0.5, p_for_reproduction=0.5` - если это не все параметры, то может уже имеет смысл выделить структуру `PopulationParams` ?

JDima commented on population_modeling/genome.py:
```diff
@@ -0,0 +1,63 @@
+        return min(parameter, 1)
+
+    def child_p_for_reproduction(self) -> float:
+        variation = norm.rvs(0, 0.01)
```
магические константы, где-то `norm.rvs(0, 0.001)`, где-то `norm.rvs(0, 0.01)` - может что-то в константы, что изменяется, чтобы потом было проще понять почему значения различаются и менять их

JDima commented on population_modeling/bacteria.py:
```diff
@@ -0,0 +1,65 @@
+            List of child's if they are, if Bacteria not alive returns None
+        """
+        if not self.is_alive:
+            raise DeadBacteria(self)
```
DeadBacteriaException

JDima commented on population_modeling/genome.py:
```diff
@@ -0,0 +1,63 @@
+
+        return child_genome
+
+    def spontaneous_mutation(self) -> None:
```
по коду не видно, что бывают альтернативные мутации
 
 не должно ли быть просто `mutate`

MaEgV committed:
Fix mypy options

JDima commented on population_modeling/bacteria.py:
```diff
@@ -0,0 +1,65 @@
+        self.is_alive = True
+        self.genome = p
+
+    def iteration(self, selector: Selector) -> list:
```
Может нам вынести этот метод из класса и сделать Bacteria константным датаклассом? Сейчас не очень понятно как действие "итерация" связана с объектом бактерия, кажется они могут жить по отдельности

JDima commented on population_modeling/genome.py:
```diff
@@ -0,0 +1,63 @@
+
+    def spontaneous_mutation(self) -> None:
+        '''
+        Spontaneous mutation of the genome
```
Все геномы будут мутировать одинаково? Может сделаем абстрактный класс Геном и от него отнаследуемся. Как альтернатива можем поступить как с бактерий - вынести мутации в отдельный класс и геном сделать константным датаклассом

JDima commented on population_modeling/population.py:
```diff
@@ -0,0 +1,118 @@
+        self.selector = selector_
+        self.graph = create_graph(create_bacteria(max_life_time, p_for_death, p_for_reproduction))
+
+    def iteration(self):
```
Кажется Популяцию можно тоже сделать константным датаклассом и вынести итерации

besperspektivnyak committed:
Fix code structure after reviewing

besperspektivnyak committed:
add documentation

documentation files made by Sphinx

MaEgV committed:
Decorative fixes in genome.py

besperspektivnyak committed:
Add new mutations file

MaEgV committed:
Integrating new mutations system in iteration

MaEgV committed:
Reorganisation of the population structure

besperspektivnyak committed:
Fix tests after reorganisation

MaEgV committed:
Documenting the code

MaEgV committed:
Documenting the code

MaEgV committed:
Merge remote-tracking branch 'origin/develop' into develop

MaEgV committed:
First part of new description

MaEgV committed:
Mypy errors fix

MaEgV committed:
Merge remote-tracking branch 'origin/develop' into develop

MaEgV committed:
Mypy errors fix

MaEgV committed:
Mypy errors fix

MaEgV committed:
Mypy errors fix

MaEgV committed:
Decorative fix

besperspektivnyak committed:
Fix comments

besperspektivnyak committed:
Fix comments

besperspektivnyak committed:
Update documentation

MaEgV committed:
Completing the readme update

Current project status:
1) The Genomes, Bacteria, Population, External Factors classes contain only the attribute initialization method(data storage);
2) Now, for each iteration by population and bacterium, you must explicitly specify instances of the Selector and MutationalProcesses classes;
3) It is possible to customize and configure the selection and mutation operators during program execution.

JDima commented on README.md:
````diff
@@ -1 +1,136 @@
+### Population
+This class stores a graph of the bacteria in the population. The graph forms a tree, so the only attribute of the class is the genealogical tree field. It is also passed to initialize an instance of the class when it is created:  
+```Python
+    def __init__(self, genealogical_tree: igraph.Graph):
````
Сомнительные вставки кода, которые не очень инофрмаивны. Может лучше пример использования - в данном случае создания. Аналогично нижние код сниппеты

JDima commented on population_modeling/bacteria.py:
```diff
@@ -0,0 +1,54 @@
+from population_modeling.selector import Genome
+
+
+class Bacteria:
```
Кажется можно тогда сделать с помощью dataclass и метод построения сделать статическим

JDima replied:
аналогично другие классы

JDima commented on population_modeling/mutations.py:
```diff
@@ -0,0 +1,173 @@
+from population_modeling.genome import Genome
+
+
+class MutationalProcesses:
```
`AbstractMutation` или `AbstractMutationProcess`

JDima commented on population_modeling/mutations.py:
```diff
@@ -0,0 +1,173 @@
+        return childs_genome
+
+
+class NormalMutations(MutationalProcesses):
```
Можно вынести в отдельный файл. Например создать папку mutations и в ней для каждого класса мутации свой файл

JDima commented on population_modeling/mutations.py:
```diff
@@ -0,0 +1,173 @@
+    """
+
+    MAX_PROBABILITY = 1
+    PARAMS_SPONTANEOUS_MUTATION = {'loc': 0, 'scale': 0.001}
```
`{'loc': 0, 'scale': 0.001}` это видимо какой тип параметров, который можно отнести в кдатакласс с двумя полями

JDima commented on population_modeling/mutations.py:
```diff
@@ -0,0 +1,173 @@
+
+    MAX_PROBABILITY = 1
+    PARAMS_SPONTANEOUS_MUTATION = {'loc': 0, 'scale': 0.001}
+    PARAMS_MAX_N_VARIATION = {'loc': 0, 'scale': 3}
```
может лучше ожидать параметры в конструкторе. чтобы можно было создавать экземпляры класса с различными параметрами

JDima commented on population_modeling/population.py:
```diff
@@ -0,0 +1,94 @@
+    return Population(genealogical_tree)
+
+
+def draw(population: Population, filename: str = None) -> None:
```
можно сделать методом популяции

JDima commented on population_modeling/bacteria_iteration.py:
```diff
@@ -0,0 +1,72 @@
+        List of child's if they are, if Bacteria not alive returns empty list
+
+    """
+    if not bacteria.is_alive:
```
получается что у бактерии наружу торчит атрибут класса, который мы чекаем, но мы можем его изменить и все сломается

JDima commented on population_modeling/bacteria_iteration.py:
```diff
@@ -0,0 +1,72 @@
+        bacteria.is_alive = False
+        return []
+
+    bacteria.age += 1
```
точно хотим иметь возможность из метода напрямую изменять атрибуты класса? или лучше сделать на уровне бактерии изменение возраста

JDima commented on population_modeling/bacteria_iteration.py:
```diff
@@ -0,0 +1,72 @@
+
+    mutation_mode.mutation(bacteria.genome)
+
+    if bacteria.age > bacteria.genome.max_life_time or selector.is_died(bacteria.genome):
```
`bacteria.age > bacteria.genome.max_life_time` - эту проверку может предоставлять класс бактерии - `check_age`

MaEgV committed:
Dataclasses adding

MaEgV committed:
Code inserts fix

besperspektivnyak committed:
Add new module 'mutations'

besperspektivnyak committed:
Add comments for new files

besperspektivnyak committed:
Change documentation style

JDima commented on README.md:
````diff
@@ -1 +1,143 @@
+## Example
+Consider a simple example of setting up a population and drawing the result of an evolution simulation:
+```Python
+from population_modeling import *
````
так лучше не импортить

MaEgV committed:
Dataclasses adding

MaEgV committed:
Merge remote-tracking branch 'origin/develop' into develop

# Conflicts:
#	README.md

MaEgV committed:
Typo fix

MaEgV committed:
Readme typos fix

MaEgV committed:
Example import fix

besperspektivnyak committed:
Update documentation

JDima commented on population_modeling/bacteria.py:
```diff
@@ -0,0 +1,112 @@
+
+        return self._status.is_alive
+
+    def inc_age(self) -> None:
```
противоречит @dataclass(frozen=True)

MaEgV committed:
Genome to dataclass

JDima commented on population_modeling/bacteria.py:
```diff
@@ -0,0 +1,112 @@
+        None
+        """
+
+        self._status.age += 1
```
в статусе

JDima commented on population_modeling/bacteria.py:
```diff
@@ -0,0 +1,112 @@
+        None
+        """
+
+        self._status.is_alive = False
```
в статусе

JDima commented on population_modeling/bacteria.py:
```diff
@@ -0,0 +1,112 @@
+            Bacteria status
+        """
+
+        return self._status.is_alive
```
в статусе

JDima commented on population_modeling/bacteria.py:
```diff
@@ -0,0 +1,112 @@
+
+
+def create_bacteria(
+            max_life_time: int = 5,
```
`GenomeParams`

JDima commented on population_modeling/mutations/normal_mutation.py:
```diff
@@ -0,0 +1,140 @@
+from population_modeling.mutations.variation_parameters import MutationalParams
+
+
+class NormalMutations(AbstractMutation):
```
NormalMutation

JDima replied:
NormalMutator

JDima commented on population_modeling/mutations/normal_mutation.py:
```diff
@@ -0,0 +1,140 @@
+    MAX_PROBABILITY = 1
+
+    def __init__(self,
+                 mutational_params: MutationalParams = MutationalParams(0, 0.01),
```
```
class NormalMutationParams(MutationParams):
     write()
     read()
```

JDima commented on population_modeling/mutations/normal_mutation.py:
```diff
@@ -0,0 +1,140 @@
+
+        return min(parameter, NormalMutations.MAX_PROBABILITY)
+
+    def mutation(self, genome: Genome) -> None:
```
mutate

JDima replied:
genome -> genome
return genome

JDima commented on population_modeling/mutations/normal_mutation.py:
```diff
@@ -0,0 +1,140 @@
+
+        """
+
+        genome.p_for_death = min(genome.p_for_death + norm.rvs(self.mutational_params.loc, self.mutational_params.scale)
```
```
p_for_death = min(genome.p_for_death + norm.rvs(self.mutational_params.loc, self.mutational_params.scale)
p_for_reproduction = min(genome.p_for_reproduction + norm.rvs(self.mutational_params.loc,
return Genome(p_for_death , p_for_reproduction )
```

JDima replied:
`def update(self, **kwargs):`

```
genome.update(p_for_death=p_for_death, p_for_reproduction=p_for_reproduction )
genome.update(p_for_death=p_for_death)
```

JDima commented on population_modeling/population.py:
```diff
@@ -0,0 +1,95 @@
+        )
+
+
+def create_graph(first_bacteria: Bacteria) -> igraph.Graph:
```
def create_graph(bacteria: Bacteria) -> igraph.Graph:

JDima replied:
create_population(bacteria: Bacteri)

JDima commented on population_modeling/population.py:
```diff
@@ -0,0 +1,95 @@
+    return graph
+
+
+def create_population(first_bacteria: Bacteria) -> Population:
```
def create_population(bacteria: Bacteria) -> Population:

JDima commented on population_modeling/population_iteration.py:
```diff
@@ -0,0 +1,120 @@
+    )
+
+
+def _update_properties(population: Population, selector: Selector) -> None:
```
не испольузется популяция

JDima commented on population_modeling/population_iteration.py:
```diff
@@ -0,0 +1,120 @@
+    """
+
+    # Add children to graph vertices with new generation labels
+    population.genealogical_tree.add_vertices(
```
перенести в популяцию

JDima commented on population_modeling/population_iteration.py:
```diff
@@ -0,0 +1,120 @@
+from population_modeling.selector import Selector
+from population_modeling.mutations.normal_mutation import AbstractMutation
+from population_modeling.bacteria_iteration import iterate as iterate_bacteria
+
```
```
class LifeCycle:
     def __init__(population):
           pass
     def iterate(selector, mutator, draw_func=lambda x: pass):
           new_generation = []
           for bacteria in population:
                   if not bacteria.is_alive():
                           raise DeadBacteriaException(bacteria)
                   mutator.mutate(bacteria)
                   if selector.is_died(bacteria.genome):
                            bacteria.die()
                            return []
                    bacteria.inc_age()
                    
                    new_generation.extend(_get_children(selector, mutation_mode, bacteria))
           population.update(new_generation)
           draw_func(population)    
```

JDima commented on population_modeling/selector.py:
```diff
@@ -0,0 +1,147 @@
+    return -ext_factors.overpopulation + genome.p_for_reproduction > uniform.rvs(0, 1)
+
+
+class Selector:
```
BaseSelector and AbstractSelector

JDima commented on population_modeling/mutations/variation_parameters.py:
```diff
@@ -0,0 +1,20 @@
+
+
+@dataclass
+class MutationalParams:
```
MutationParams

besperspektivnyak committed:
Add methods from Bacteria to BacteriaStatus

MaEgV committed:
Add life cycle + selector restructuring

MaEgV committed:
Mypy fix

MaEgV committed:
Mypy fix

besperspektivnyak committed:
Add upload method to genome.py

besperspektivnyak committed:
Fix types

besperspektivnyak committed:
Update documentation
